### PR TITLE
Use a forgettable Drop impl to abort on panics

### PR DESCRIPTION
**Pull request by Lej77**
Its probably better to use a forgettable zero sized type with a custom `Drop` implementation to abort on panic instead of using `std::panic::catch_unwind`.

I think I first saw this pattern described at [dtolnay/case-studies/function-epilogue](https://github.com/dtolnay/case-studies/blob/master/function-epilogue/README.md).

This pattern works on `no_std` and is also more future proof in case Rust ever supports unwinds that can't be caught (maybe from `c++`?) as this pattern should abort on those unwinds as well.

### CODE REVIEW

File: scoped_async_spawn/src/guard.rs [hunk 1]
```diff
@@ -87,15 +87,19 @@ impl<'s> SpawnGuard<'s> {
...
-                    ch.as_ref().drop_future_now();
-                }
-            })
-        }));
-        if res.is_err() {
-            std::process::abort();
+        struct AbortBomb;
+        impl Drop for AbortBomb {
+            fn drop(&mut self) {
+                let _bomb = AbortBomb;
```
Comment (wishawa) [inline scoped_async_spawn/src/guard.rs:93]: Do we need this line?

**Comment by wishawa** [comment]
I like this trick. One question: is the panic on unwind = abort behavior guaranteed? I read about it somewhere but can't remember if it was official documentation or not.

File: scoped_async_spawn/src/guard.rs (hunk 1 above)
Comment (Lej77) [reply scoped_async_spawn/src/guard.rs:93]: Not really, it should never execute as long as panicking while a panic is already in progress leads to an abort. So this is really to guard against that changing. If it doesn't cause an abort we will instead call this nested abort bomb's drop implementation and panic again.

**Comment by Lej77** [comment]
I am not sure actually! That is why I added that nested `let _bomb = AbortBomb;` statement, to cause infinite recursion if it somehow doesn't lead to an abort. Still I can't think of any reason they would change that. From what I understand there has been thoughts about making all panics in Drop implementations lead to aborts. So it seems likely that if there was a change it would be to more aggressively abort.

If you want to be safer and you are using `std` anyways, then just call the `std::process::abort` function inside the `AbortBomb`'s `Drop` implementation.

**Comment by wishawa** [comment]
Awesome. Since it's abort or infinite recursion anyway I don't think we need to use `std::process::abort`.

**Pull request merged**